### ISSUE TITLE
Review firewall rules

### ISSUE THREAD

**Issue by michaelshmitty** [michaelshmitty/esperpass#7]
We seem to be missing out on some hits because the [firewall whitelist of Streetpass relay IP addresses](https://github.com/michaelshmitty/esperpass/blob/master/user/config_flash.c#L94) may be incomplete. This list is based on the [code of the Raspipass project](https://github.com/Pinchie/RaspiPass/blob/master/raspi_secure/firewall.rules).
You can temporarily turn off the firewall by issuing
```
CMD> acl from_sta clear
CMD> save
CMD> reset
```

**Note that this will leave the "attwifi" network completely open and give anyone connected to it unrestricted access to your Internet connection. Use with caution and only for testing.**

To restore the default firewall rules perform a factory reset by issuing the following commands:
```
CMD> reset factory
```
Esperpass will restart and you need to reconfigure your WiFi ssid and password.

We'll need to debug the firewall rules and see which IP addresses the 3DS connects to that are not in the list. This debugging functionality seems to be built into the original ESP WiFi Repeater so it just takes some time to configure it and sort this out.
If anyone else has time to debug this, please report your findings to this issue.

**Comment by Dal78** [comment]
It is worth noting that it might not be adresses that are missing but protocols.  Does the 3ds maybe use a UDP transmission as well as TCP?  Is the ACL just allowing TCP or all protocols?

Sometimes ICMP is also needed for a system to be happy it can see the other end.

If we can verbosely log the rejected requests from a valid 3ds mac address on a test bed you should be able to see what is triggering  the failure.

**Comment by michaelshmitty** [comment]
The ACLs currently built in are set to "IP". According to [Martin Ger's documentation](https://github.com/martin-ger/esp_wifi_repeater#firewall) that means both UDP and TCP are allowed.
I have had tags with the firewall on, just not that many. It's possible that it's just the list of actual IP addresses that is out of date / inadequate.

**Comment by Dal78** [comment]
Im not familar with coding that must just some scripting.  Where are the ACL's set in this projects source code?

**Comment by michaelshmitty** [comment]
> Im not familar with coding that much just some scripting. Where are the ACL's set in this projects source code?

I'm a novice C programmer and this project is helping me to get some more experience in this areaa.

The ACL's are set in config_flash.c in a function that reloads the config when you flash the firmware for the first time or do a factory reset:
https://github.com/michaelshmitty/esperpass/blob/master/user/config_flash.c#L94
The list I've used is based on the Raspipass project:
https://github.com/Pinchie/RaspiPass/blob/master/raspi_secure/firewall.rules

**Comment by Dal78** [comment]
You need to allow DHCP inside the network:

acl from_sta UDP any any any 67 allow

or maybe fine tune that to be only inside the local subnet if you so desire.

**Comment by michaelshmitty** [comment]
TLDR; DHCP is already enabled on the internal network. There are no acl rules configured on that interface and the acl implementation then defaults to "allow any any".

As I understand it from [the WiFi Repeater documentation](https://github.com/martin-ger/esp_wifi_repeater#firewall), from_sta is the chain that controls outbound connections from the sta (the interface connected to the Internet). So that's where the connections are restricted with the built-in firewall rules for the Streepass relays. Everything else, all other chains like inbound on the interface connected and inbound and outbound on the internal interface ("attwifi") are completely open because the acls default to "allow any any". So DHCP should already work with the firewall rules on. In fact this seems to be the case as I have no problem getting an IP address from the "attwifi" network (with the default firewall rules in place).

Again I think the issue is just that not all Streepass relays are supported in the current list and we just need to amend /  update it.

**Comment by Dal78** [comment]
"set acl_debug 1"
shows you denied requests.  I have two IP's not on the list already and did see dhcp listed on the internal network.  I agree with what you say about DHCP, but i have seen funny things before with it.  My phone couldnt pull an IP until i allowed it but i was changing the rules around a lot i might have upset something.

missing IP's so far:

52.24.183.161
52.39.19.157

**Comment by michaelshmitty** [comment]
This is great work, @Dal78 
I got one too:
23.2.226.76
I'll keep monitoring your message and in the mean time I'm working on an updated release 0.0.6 that has these new addresses in them.
I have a good feeling about this. I've just whitelisted the IP addresses you gave me and I already get hits on my 3DS with the firewalled esperpass.
So the Raspipass list is definitely out of date.
We don't currently seem to be having space issues but one could imagine that adding more and more acls for every IP could potentially start causing memory and/or processing issues on the ESP8266. It would be nice to look at the output of `show acl` from time to time and note the IP addresses that never get a hit. We could then remove them from the list altogether and keep things nice and clean.

The list of MAC addresses could be next. There are so many on [homepass.info](http://homepass.info), it's making my head spin. I've just added those few "official" MAC addresses but I cannot help but think that this may limit the amount of supported games that still use Streetpass and use one of these "special" mac addresses. Anyway first things first let's get this IP list updated. Thanks a lot for your help!

**Comment by Dal78** [comment]
Thats no worries.  I fear this list might be dynamic.  The 3ds could be putting in a query and getting dynamic AWS addresses which it then connects to for the actual hits.

**Comment by michaelshmitty** [comment]
Ah yes, a dynamic list would be hard to maintain. Unfortunately the firewall implementation doesn't support MAC address filtering. Otherwise you could then get the mac address from the 3DS system settings and whitelist that in esperpass. Although I'm not sure that would add a lot of security because mac addresses can be easily spoofed and someone could probably sniff your 3ds mac address from the unencrypted "attwifi" network and still gain full internet access.

Another way would be to disable the firewall and clearly warn that using esperpass bridges your internet connection to an unprotected wifi network. Then it would be up to the user to just manually switch on the ESP8266 from time to time to get the tags.
Kind of breaks the "set and forget" flow though...

Or we could build in a timer that would bring up the "attwifi" network for a shorter time at random intervals. I think Raspipass does this to reduce the number of streetpass hits and better simulate a situation where you would actually be walking on the street randomly encountering relays.

Let's think about this a bit. I'll release an updated version with the new IP addresses anyway and we'll take it from there. If the list is indeed (very) dynamic then we'll figure it out soon enough.

**Comment by Dal78** [comment]
I think the best solution might be a random mac address from the pool every 15 minutes opening the AP external wifi for 90 seconds and then closing it until the next rotation.

Hits would be quite natural like that and nobody would be able to make much use of the AP for anything other than 3ds hits.  You could also test if DNS is required for the 3ds to get streetpass hits, maybe not.  If so turning DNS off reduces the attractiveness of the AP quite heavily for freeloaders even if the firewall is not restricted.

Did you see the Nintendo notice of their European shutdown?  We might not have long left before the shut down the AWS.

https://www.nintendo.co.uk/Nintendo-3DS-Family/StreetPass/Nintendo-Zone-Finder/Nintendo-Zone-Finder-116574.html

**Comment by michaelshmitty** [comment]
@Dal78 Good suggestions, sound approach. I'll see if it's worth working on because with this European shutdown notice following the earlier Japanese notice there is indeed little hope for Streetpass to keep working. Although I think the Nintendo zone did more than just Streetpass, so that's definitely some relays that will disappear. But they may still keep the infrastructure up to support the relays at other places like game shops and such.

In conclusion, the shutdown of Nintendo zones may not spell doom for Streetpass just yet.

**Comment by michaelshmitty** [comment]
@Dal78 I went ahead and implemented this. It's now running at my home and so far it seems to be working well.
Here's more info in the release notes:
https://github.com/michaelshmitty/esperpass/releases/tag/v0.1.0

I tried disabling DNS but it appeared that tags were not coming through. I suspect DNS is required by StreetPass unfortunately. However with the AP now shutting down after 90 seconds there is indeed a smaller chance that freeloaders will find this useful.
Next it may be interesting to look at the MAC address list.

Thanks for the suggestions! Closing this issue.

**Comment by Dal78** [comment]
Nice work I'll flash it on now :)
